In [ ]:
import numpy as np 
import pandas as pd 
import seaborn as sns 
from scipy.stats import t,ttest_1samp,ttest_ind
from scipy.stats import t , skewnorm , skew , norm , ttest_1samp , ttest_ind
import matplotlib.ticker as ticker
from sklearn import datasets, linear_model
from sklearn.decomposition import PCA
from sklearn.preprocessing import MinMaxScaler
from sklearn.manifold import TSNE

from numpy.random import seed
from matplotlib import pyplot as plt
%matplotlib inline
sns.set_style("whitegrid")

In [ ]:
df = pd.read_csv('SpotifyFeatures.csv')
df.head()

# in the above two cells we've imported the data .
## from the data we can see that dataset consists of songs along with
## various attributes like duration,energy ,key, liveness etc.

In [12]:
pd.isnull(df).sum()


genre               0
artist_name         0
track_name          0
track_id            0
popularity          0
acousticness        0
danceability        0
duration_ms         0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

In [15]:
np.percentile(df['popularity'],90)

63.0

### we can say that any points outside the range [lower_bound,upper_bound] are outliers
### interestingly in outliers also points such lower tempos (i.e slower songs)
### have significantly lesser popularity than points with higher tempos 
### so we can say that songs with higher tempos are more likely to be popular 

## Feature Engineering ##

In [3]:
df.dtypes

genre                object
artist_name          object
track_name           object
track_id             object
popularity            int64
acousticness        float64
danceability        float64
duration_ms           int64
energy              float64
instrumentalness    float64
key                  object
liveness            float64
loudness            float64
mode                 object
speechiness         float64
tempo               float64
time_signature       object
valence             float64
dtype: object

In [4]:
df.select_dtypes(include=['object']).columns

Index(['genre', 'artist_name', 'track_name', 'track_id', 'key', 'mode',
       'time_signature'],
      dtype='object')

In [5]:
# apart from the artist_name,track_name,track_id we have 3 catagorical variables which are
# - key 
# - mode
# - time_signature

In [6]:
list_of_keys = df['key'].unique()

In [7]:
len(df['key'].unique())

12

In [8]:
for i in range(len(df['key'].unique())):
    df.loc[df['key'] == list_of_keys[i], 'key'] = i
    

In [9]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness  key  liveness  loudness   mode  \
0        99373   0.910             0.000    0    0.3460    -1.828  Major   
1       137373   0.737             0.000    1    0.1510    -5.559  Minor   
2       170267   0.131             0.000    2    0.1030   -13.879  Minor   
3       152427   0.326             0.000    0    0.0985   -12.178  Major   
4        82625   0.225             0.123    3    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [10]:
df.loc[df['mode'] == 'Major', 'mode'] = 1
df.loc[df['mode'] == 'Minor', 'mode'] = 0



In [11]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness  key  liveness  loudness  mode  \
0        99373   0.910             0.000    0    0.3460    -1.828     1   
1       137373   0.737             0.000    1    0.1510    -5.559     0   
2       170267   0.131             0.000    2    0.1030   -13.879     0   
3       152427   0.326             0.000    0    0.0985   -12.178     1   
4        82625   0.225             0.123    3    0.2020   -21.150     1   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [12]:
time_signatures = df['time_signature'].unique()

In [13]:
for i in range(len(time_signatures)):
    df.loc[df['time_signature']==time_signatures[i],'time_signature'] = i

In [14]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness  key  liveness  loudness  mode  \
0        99373   0.910             0.000    0    0.3460    -1.828     1   
1       137373   0.737             0.000    1    0.1510    -5.559     0   
2       170267   0.131             0.000    2    0.1030   -13.879     0   
3       152427   0.326             0.000    0    0.0985   -12.178     1   
4        82625   0.225             0.123    3    0.2020   -21.150     1   

   speechiness    tempo  time_signature  valence  
0       0.0525  166.969               0    0.814  
1       0.0868  174.003               0    0.816  
2       0.0362   99.488               1    0.368  
3       0.0395  171.758               0    0.227  
4       0.0456  140.576               0    0.390

## set the threshold for popularity##

In [1]:
# Definition of Popularity 
# we define a song to be popular if it is 
# more popular than 90 percentile of popularity

In [16]:
np.percentile(df['popularity'],90)

63.0

In [17]:
# so in our case we define a song to be popular if its popularity is greater than 63 else not 
# this is turning into a classification problem 


## Machine Learning ##

In [18]:
df.loc[df['popularity'] < 63, 'popularity'] = 0 
df.loc[df['popularity'] >= 63, 'popularity'] = 1

In [19]:
df.loc[df['popularity'] == 1]


genre                artist_name  \
135      R&B              Mary J. Blige   
136      R&B                    Rihanna   
139      R&B             Olivia O'Brien   
141      R&B                        Nao   
142      R&B                Frank Ocean   
144      R&B                      Usher   
145      R&B               Mariah Carey   
146      R&B                    Beyoncé   
147      R&B               Jason Derulo   
149      R&B                   Big Sean   
150      R&B         Christina Aguilera   
153      R&B                      6LACK   
154      R&B                        SZA   
155      R&B                  Kyle Dion   
156      R&B                  Roy Woods   
157      R&B                    Tove Lo   
158      R&B                James Blake   
159      R&B                       Joji   
160      R&B             Jennifer Lopez   
161      R&B                    Rihanna   
162      R&B                     Miguel   
163      R&B                        TLC   
164      R&B              PARTYNEXTDOOR   
165      R&B                 MIKExANGEL   
166      R&B                 Nick Jonas   
167      R&B                 Tory Lanez   
168      R&B                Chris Brown   
172      R&B                  PJ Morton   
173      R&B                  Tom Misch   
174      R&B                      Usher   
...      ...                        ...   
226250  Rock              Elvis Presley   
226259  Rock                       Maná   
226281  Rock             Alejandro Sanz   
226284  Rock                 Pretenders   
226298  Rock                 Papa Roach   
226302  Rock                  Kate Bush   
226303  Rock                  Pearl Jam   
226307  Rock                   Elefante   
226326  Rock                       Maná   
226333  Rock            Jesse McCartney   
226345  Rock           System Of A Down   
226351  Rock                Snow Patrol   
226354  Rock                    The Jam   
226400  Rock                     Travis   
226401  Rock        Heroes Del Silencio   
226405  Rock                   Bee Gees   
226412  Rock        Secondhand Serenade   
226413  Rock                    Roxette   
226416  Rock             George Michael   
226421  Rock             Franco De Vita   
226424  Rock           La Mosca Tse-Tse   
226438  Rock           System Of A Down   
226453  Rock  Nick Cave & The Bad Seeds   
226484  Rock               Die Antwoord   
226490  Rock                  Sam Cooke   
226617  Soul                Craig David   
226669  Soul              George Benson   
227075  Soul                Barry White   
227951  Soul             Randy Crawford   
228607  Soul                Seinabo Sey   

                                               track_name  \
135                            Be Without You - Kendu Mix   
136                                             Desperado   
139                                           Love Myself   
141                         Make It Out Alive (feat. SiR)   
142                                             Seigfried   
144                                  You Make Me Wanna...   
145                                                  Hero   
146                                        Flawless Remix   
147                        Tip Toe (feat. French Montana)   
149                                                I Know   
150       Lady Marmalade - From "Moulin Rouge" Soundtrack   
153                                            Disconnect   
154                 Childs Play (feat. Chance the Rapper)   
155                                                 Brown   
156                                          Get You Good   
157                                    Habits (Stay High)   
158                         Can't Believe The Way We Flow   
159          SLOW DANCING IN THE DARK - Loud Luxury Remix   
160                                                Dinero   
161                                  Hate That I Love You   
162                                     Remember Me (Dúo)  

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier


from sklearn.metrics import make_scorer, accuracy_score, roc_auc_score 
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import train_test_split

In [ ]:
# pip install xgboost

In [ ]:
features = ["acousticness", "danceability", "duration_ms", "energy", "instrumentalness", "key", "liveness", 
            "mode", "speechiness", "tempo", "time_signature", "valence"]

In [ ]:
training = df.sample(frac = 0.8,random_state = 420)
X_train = training[features]
y_train = training['popularity']
X_test = df.drop(training.index)[features]

In [ ]:
X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size = 0.2, random_state = 420)

In [34]:
LR_Model = LogisticRegression()
LR_Model.fit(X_train, y_train)
LR_Predict = LR_Model.predict(X_valid)
LR_Accuracy = accuracy_score(y_valid, LR_Predict)
print("Accuracy: " + str(LR_Accuracy))

LR_AUC = roc_auc_score(y_valid, LR_Predict) 
print("AUC: " + str(LR_AUC*100))

c:\users\ravi\appdata\local\programs\python\python36\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accuracy: 0.8895692340745515
AUC: 50.0


In [35]:
KNN_Model = KNeighborsClassifier()
KNN_Model.fit(X_train, y_train)
KNN_Predict = KNN_Model.predict(X_valid)
KNN_Accuracy = accuracy_score(y_valid, KNN_Predict)
print("Accuracy: " + str(KNN_Accuracy))

KNN_AUC = roc_auc_score(y_valid, KNN_Predict) 
print("AUC: " + str(KNN_AUC*100))

Accuracy: 0.8747985820174025
AUC: 59.234086855334766


In [36]:
DT_Model = DecisionTreeClassifier()
DT_Model.fit(X_train, y_train)
DT_Predict = DT_Model.predict(X_valid)
DT_Accuracy = accuracy_score(y_valid, DT_Predict)
print("Accuracy: " + str(DT_Accuracy))

DT_AUC = roc_auc_score(y_valid, DT_Predict) 
print("AUC: " + str(DT_AUC))

Accuracy: 0.9157804275432377
AUC: 0.8334886227532248


In [37]:
RFC_Model = RandomForestClassifier()
RFC_Model.fit(X_train, y_train)
RFC_Predict = RFC_Model.predict(X_valid)
RFC_Accuracy = accuracy_score(y_valid, RFC_Predict)
print("Accuracy: " + str(RFC_Accuracy))

RFC_AUC = roc_auc_score(y_valid, RFC_Predict) 
print("AUC: " + str(RFC_AUC))

c:\users\ravi\appdata\local\programs\python\python36\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


Accuracy: 0.9542109786228381
AUC: 0.8122760596076434


In [38]:

model_performance_accuracy = pd.DataFrame({'Model': ['LogisticRegression', 
                                                      'RandomForestClassifier', 
                                                      'KNeighborsClassifier',
                                                      'DecisionTreeClassifier'
                                                    ],
                                            'Accuracy': [LR_Accuracy,
                                                         RFC_Accuracy,
                                                         KNN_Accuracy,
                                                         DT_Accuracy
                                                         ]})

model_performance_AUC = pd.DataFrame({'Model': ['LogisticRegression', 
                                                      'RandomForestClassifier', 
                                                      'KNeighborsClassifier',
                                                      'DecisionTreeClassifier'],
                                            'AUC': [LR_AUC,
                                                         RFC_AUC,
                                                         KNN_AUC,
                                                         DT_AUC]})

In [39]:
model_performance_accuracy.sort_values(by = "Accuracy", ascending = False)


Model  Accuracy
1  RandomForestClassifier  0.954211
3  DecisionTreeClassifier  0.915780
0      LogisticRegression  0.889569
2    KNeighborsClassifier  0.874799

### random forest got the most accuracy , lets see if we can optimize it further using hyperparameter tuning 

In [44]:
from sklearn.model_selection import RandomizedSearchCV
import pprint # for pretty printing


# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

pprint.pprint(random_grid)

{'bootstrap': [True, False],
 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}


In [50]:

improved_model = RandomForestClassifier(bootstrap = True,
 max_depth = 70,
 max_features =  'auto',
 min_samples_leaf= 4,
 min_samples_split =  10,
 n_estimators = 400)

In [51]:
improved_model.fit(X_train, y_train)
RFC_Predict = improved_model.predict(X_valid)
RFC_Accuracy = accuracy_score(y_valid, RFC_Predict)
print("Accuracy: " + str(RFC_Accuracy))

RFC_AUC = roc_auc_score(y_valid, RFC_Predict) 
print("AUC: " + str(RFC_AUC))

Accuracy: 0.9200236330432915
AUC: 0.6456636441443707


In [53]:
improved_model.fit(X_train, y_train)
RFC_Predict = improved_model.predict(X_test)
RFC_Accuracy = accuracy_score(y_test, RFC_Predict)
print("Accuracy: " + str(RFC_Accuracy))

RFC_AUC = roc_auc_score(y_test, RFC_Predict) 
print("AUC: " + str(RFC_AUC))

NameError: name 'y_test' is not defined

### strange .. after hyperparameter tuning the accuracy went down from 95% to 92%